src/mirage_kv_unix.ml

```diff
@@ -46,224 +71,298 @@ let split_string delimiter name =
-      | Ok fd ->
-        Lwt.catch
-          (fun () ->
-             let buf = Bytes.unsafe_of_string value in
-             let rec write_once off len =
-               if len = 0 then Lwt_unix.close fd
-               else
-                 Lwt_unix.write fd buf off len >>= fun n_written ->
-                 if n_written = len + off then
-                   Lwt_unix.close fd
-                 else
-                   write_once (off + n_written) (len - n_written)
-             in
-             write_once 0 (String.length value) >|= fun () ->
-             Ok ())
-          (function
-            | Unix.Unix_error (Unix.ENOSPC, _, _) -> Lwt.return (Error `No_space)
-            | Unix.Unix_error (e, _, _) -> Lwt.return (Error (`Storage_error (key, Unix.error_message e)))
-            | e -> Lwt.fail e)
+let set_aux t key ?offset value =
+  let lseek fd =
+    match offset with
+    | None -> Lwt.return ()
+    | Some offset ->
+        let+ _ =
+          Lwt_unix.LargeFile.lseek fd (Optint.Int63.to_int64 offset) SEEK_SET
+        in
+        ()
+  in
+  let* exists = exists t key in
+  match exists with
+  | Ok (Some `Dictionary) -> Lwt.return (err_key_exists key)
+  | _ -> (
+      let* ret =
+        match offset with
+        | None -> remove t key (* set always overwite the previous bindings. *)
+        | Some _ -> Lwt.return (Ok ())
+      in
+      match ret with
+      | (Error (`Dictionary_expected _) | Error (`Storage_error _)) as e ->
+          Lwt.return e
+      | Ok () | Error (`Not_found _) ->
+          let*? fd =
+            open_file t key Lwt_unix.[ O_WRONLY; O_NONBLOCK; O_CREAT ]
+          in
+          Lwt.catch
+            (fun () ->
+              let* () = lseek fd in
+              let buf = Bytes.unsafe_of_string value in
+              let rec write_once off len =
+                if len = 0 then Lwt_unix.close fd
+                else
+                  let* n_written = Lwt_unix.write fd buf off len in
+                  if n_written = len + off then Lwt_unix.close fd
+                  else write_once (off + n_written) (len - n_written)
+              in
+              let+ () = write_once 0 (String.length value) in
+              Ok ())
+            (function
+              | Unix.Unix_error (ENOSPC, _, _) -> Lwt.return err_no_space
+              | Unix.Unix_error (e, _, _) -> Lwt.return (err_unix key e)
+              | e -> Lwt.fail e))
+
+let set_partial t key ~offset value = set_aux t key ~offset value
+let set t key value = set_aux t key value
+
+let allocate t key ?last_modified length =
+  let len = Int63.to_int length in
+  let value = String.make len '\000' in
+  let set_last_modified () =
+    match last_modified with
+    | None -> Lwt.return (Ok ())
+    | Some ts ->
+        let path = resolve_filename t.base key in
+        let date = Ptime.to_float_s ts in
+        let+ () = Lwt_unix.utimes path date date in
+        Ok ()
+  in
+  let*? () = set t key value in
+  set_last_modified ()
```

Comment:
`Lwt_unix.ftruncate`?

Reply:
As we need to make sure the error path is similar to `set` so probably easier to keep `set` here?